spring-kafka/src/main/java/org/springframework/kafka/support/DefaultKafkaHeaderMapper.java

```diff
@@ -324,12 +313,42 @@ else if (!(headerName.equals(JSON_TYPES)) && matchesForInbound(headerName)) {
 					populateJsonValueHeader(header, requestedType, headers);
 				}
 				else {
-					headers.put(headerName, headerValueToAddIn(header));
+					fromUserHeader(headerName, header, headers);
 				}
 			}
 		});
 	}
 
+	private void fromHeader(String key, Object rawValue, Map<String, String> jsonHeaders,
+							ObjectMapper headerObjectMapper, Headers target) {
+
+		Object valueToAdd = headerValueToAddOut(key, rawValue);
+		if (valueToAdd instanceof byte[]) {
+			target.add(new RecordHeader(key, (byte[]) valueToAdd));
+		}
+		else {
+			try {
+				String className = valueToAdd.getClass().getName();
+				boolean encodeToJson = this.encodeStrings;
+				if (this.toStringClasses.contains(className)) {
+					valueToAdd = valueToAdd.toString();
+					className = JAVA_LANG_STRING;
+					encodeToJson = true;
+				}
+				if (!encodeToJson && valueToAdd instanceof String) {
+					target.add(new RecordHeader(key, ((String) valueToAdd).getBytes(getCharset())));
+				}
+				else {
+					target.add(new RecordHeader(key, headerObjectMapper.writeValueAsBytes(valueToAdd)));
+				}
+				jsonHeaders.put(key, className);
```

Comment:
So, if original header is an iterable we call this `fromHeader` that many times.
While `target.add(new RecordHeader())` is OK , because Kafka supports multi-value headers. but this `jsonHeaders` is a single-value map, so we override the entry every time we iterate.
Might be OK to have only one type per iterabel, but can we optimize it a little bit with a `putIfAbsent()`?
And probably document somewhere that all the items of the iterable header must be of the same type.